File: frontend/src/types/node.ts
// Node Types
// Mirrors: project/src/daemon/types/node.rs

import type { TransactionViewResponse } from './transaction';

export interface NodeStatusResponse {
  version: string;
  peers_connected: number;
  block_height: number;
  top_block_hash: string;
  /** Hash of the local genesis block; identifies the network. Empty (or all-zero)
   *  when the chain is empty. */
  genesis_hash: string;
}

export interface MempoolEntry {
  tx: TransactionViewResponse;
}

export interface MempoolResponse {
  count: number;
  transactions: MempoolEntry[];
}

export interface NewPeerConnectionResponse {
  success: boolean;
  fail_message: string | null;
}

File: project/src/daemon/types/node.rs
// Node RPC Types
use serde::{Deserialize, Serialize};

use crate::daemon::types::TransactionViewResponse;

#[derive(Serialize, Deserialize, Debug)]
pub struct NodeStatusResponse {
    pub version: String,
    pub peers_connected: usize,
    pub advertised_addr: String,
    pub block_height: usize,
    pub top_block_hash: String,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct MempoolEntry {
    pub tx: TransactionViewResponse,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct MempoolResponse {
    pub count: usize,
    pub transactions: Vec<MempoolEntry>,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct NodeInitResponse {
    pub success: bool,
    pub block_count: usize,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct SimpleSuccessResponse {
    pub success: bool,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct NewPeerConnectionParams {
    pub address: String,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct NewPeerConnectionResponse {
    pub success: bool,
    pub fail_message: Option<String>,
}
